File: ts/utils/map.ts
declare var mapFunctions: { [id: string]: () => void };
mapFunctions = {};

declare var _G: any;
_G['callmapFunction'] = (id: string) => {
  mapFunctions[id]();
};

type OptionsType = {
  mode: 'n' | 'i' | 't';
  noremap?: boolean;
  silent?: boolean;
  expr?: boolean;
};
export const map = (id: string, key: string, options: OptionsType, command: string | (() => void)) => {
  const { mode, ...opts } = options;
  // prettier-ignore
  const commandText = 
    typeof command === 'string' 
      ? command 
      : `<C-O>:lua callmapFunction("${id}")<CR>`;
  if (typeof command !== 'string') {
    mapFunctions[id] = command;
  }
  vim.api.nvim_set_keymap(mode, key, commandText, opts);
};

File: lua/utils/map.lua
return {
  nmap = function(key, action, options)
    vim.api.nvim_set_keymap("n", key, action, options)
  end,
  imap = function(key, action, options)
    vim.api.nvim_set_keymap("i", key, action, options)
  end,
  tmap = function(key, action, options)
    vim.api.nvim_set_keymap("t", key, action, options)
  end,
  vmap = function(key, action, options)
    vim.api.nvim_set_keymap("", key, action, options)
  end,
  map = function(key, action, options)
    vim.api.nvim_set_keymap("", key, action, options)
  end,
  n = { noremap = true },
  ns = { noremap = true, silent = true },
  s = { silent = true },
  get_buf_map = function(bufnr, mode)
    return function(key, action, options)
      vim.api.nvim_buf_set_keymap(bufnr, mode, key, action, options)
    end
  end,
}
